src/probability_theory/martingale.lean

```diff
@@ -0,0 +1,92 @@
+/-
+Copyright (c) 2021 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne
+-/
+
+import probability_theory.stopping
+import measure_theory.function.conditional_expectation
+
+/-!
+# Martingales
+
+A family of functions `f : ι → α → E` is a martingale with respect to a filtration `ℱ` if every
+`f i` is integrable, `f` is adapted with respect to `ℱ` and for all `i ≤ j`,
+`μ[f j | ℱ.le i] =ᵐ[μ] f i`.
+The definitions of filtration and adapted can be found in `probability_theory.stopping`.
+
+### Definitions
+
+* `is_martingale f ℱ μ`: `f` is a martingale with respect to filtration `ℱ` and measure `μ`.
+
+### Results
+
+* `is_martingale_condexp f ℱ μ`: the sequence `λ i, μ[f | ℱ i, ℱ.le i])` is a martingale with
+  respect to `ℱ` and `μ`.
+
+-/
+
+open topological_space filter
+open_locale nnreal ennreal measure_theory
+
+namespace measure_theory
+
+variables {α E ι : Type*} [preorder ι] [measurable_space E]
+  {m0 : measurable_space α} {μ : measure α}
+  [normed_group E] [normed_space ℝ E] [complete_space E] [borel_space E]
+  [second_countable_topology E]
+  {f g : ι → α → E} {ℱ : filtration ι m0} [sigma_finite_filtration μ ℱ]
+
+/-- A family of functions `f : ι → α → E` is a martingale with respect to a filtration `ℱ` if every
+`f i` is integrable, `f` is adapted with respect to `ℱ` and for all `i ≤ j`,
+`μ[f j | ℱ.le i] =ᵐ[μ] f i`. -/
+def is_martingale (f : ι → α → E) (ℱ : filtration ι m0) (μ : measure α)
+  [sigma_finite_filtration μ ℱ] : Prop :=
```

Comment:
I have a different definition of martingales without requiring `sigma_finite_filtration μ ℱ` here:
https://github.com/JasonKYi/probability_theory/blob/main/src/martingale.lean#L69
```lean
def is_martingale (μ : measure α) (f : filtration α ι m) (u : ι → Lp E 1 μ) : Prop :=
∀ ⦃i j : ι⦄, i ≤ j → ∀ ⦃s : set α⦄, measurable_set[f i] s →
  ∫ x in s, u i x ∂μ = ∫ x in s, u j x ∂μ
```
I wonder if its better without requiring it.

Reply:
You could also weaken that definition to require the equality only on measurable sets with finite measure.

A general issue with definitions like `is_martingale` is that if we require too much conditions, then some theorems apply only the restricted objects that we get, while they could be more applicable. The opposite issue is that if the definition is not restrictive enough, then we will need many additional properties to be able to prove interesting results. It could be that mine is too strong, but yours looks too weak: at least you should require that the process is adapted to the filtration.

If you require adapted + equality of integrals on finite measurable sets, you get basically the properties of the conditional expectation, minus integrability. And if I remember correctly the problems I had when doing it, obtaining an integrable version of the conditional expectation is the main reason behind the sigma finite hypotheses in the conditional expectation construction.

My instinct is to stick to the mainstream definition of a martingale: an integrable adapted process with the conditional expectation equality. If we start probability theory with general but weak definitions, I fear that we will progress very slowly. Right now we don't have much probability, so I'd say that we should first prove many things about standard objects, then maybe generalize later to more exotic settings (and in probability theory, the measure is finite anyway).

Reply:
But now that we talk about it, I realize that I can deduce integrable and adapted from the conditional expectation equality, instead of requiring separate proofs... I'll change that.

Reply:
Great! In that case my suggestion is no longer more general since it required integrability

Reply:
I think your suggestion is still more general. Your def requires integrable, should require measurability and asks for the equality of integrals. My def is basically the same but requires also a sigma_finite property. The only use of that sigma_finite property is to get existence of at least one function that verifies those properties (the conditional expectation), and then to require the martingale to be equal to that function.

Let's look at the difference on an example, where we take a function `f` and a filtration `F` and want to form a martingale from that (basically the one described in my `is_martingale_condexp` lemma): with your definition without sigma_finite, we don't have the guarantee that such a martingale exists, since it basically asks for existence of conditional expectations and they might not exist. See Example 2.6.24 in "Analysis in Banach Spaces (volume 1)", Hytönen et al. But they might also exist and we then have a martingale in your definition (not in mine, since we don't have sigma_finite).

It is not a big difference but your definition is slightly more general. For the sake of simplicity, I think we should assume `sigma_finite_filtration` and use conditional expectations.